src/core/utils/components/VirtualScroll.tsx

```diff
@@ -100,145 +131,232 @@ export type VirtualScrollProps = {
 export const VirtualScroll = React.forwardRef<
   HTMLDivElement,
   VirtualScrollProps
->(({ itemsLength, itemRenderer, bufferSize = 10, style, ...rest }, ref) => {
-  const [startNode, setStartNode] = React.useState(0);
-  const [visibleNodeCount, setVisibleNodeCount] = React.useState(0);
-  const scrollContainer = React.useRef<HTMLElement>();
-  const parentRef = React.useRef<HTMLDivElement>(null);
-  const childHeight = React.useRef(0);
-  const onScrollRef = React.useRef<(e: Event) => void>();
-  // Used only to recalculate on resize
-  const [scrollContainerHeight, setScrollContainerHeight] = React.useState(0);
-
-  const onResize = React.useCallback(({ height }) => {
-    setScrollContainerHeight(height);
-  }, []);
-  const [resizeRef] = useResizeObserver(onResize);
-
-  // Find scrollable parent
-  // Needed only on init
-  React.useLayoutEffect(() => {
-    const scrollableParent = getScrollableParent(
-      parentRef.current,
-      parentRef.current?.ownerDocument,
-    );
-    scrollContainer.current = scrollableParent;
+>(
+  (
+    {
+      itemsLength,
+      itemRenderer,
+      bufferSize = 10,
+      scrollToIndex,
+      style,
+      as = 'div',
+      parentProps,
+      ...rest
+    },
+    ref,
+  ) => {
+    const [startNode, setStartNode] = React.useState(0);
+    const [visibleNodeCount, setVisibleNodeCount] = React.useState(0);
+    const scrollContainer = React.useRef<HTMLElement>();
+    const parentRef = React.useRef<HTMLDivElement>(null);
+    const childHeight = React.useRef({ first: 0, middle: 0, last: 0 });
+    const onScrollRef = React.useRef<(e: Event) => void>();
+    // Used only to recalculate on resize
+    const [scrollContainerHeight, setScrollContainerHeight] = React.useState(0);
+    const visibleIndex = React.useRef({ start: 0, end: 0 });
+    // Used to mark when scroll container has height (updated by resize observer)
+    // because before that calculations are not right
+    const [isMounted, setIsMounted] = React.useState(false);
```

Comment:
Shouldn't this be a ref? Or does it take too long to update and you need it to re-render later?

Reply:
I need my useEffects to run, when this changes, so cannot be a ref..